File: Backend-Python/app/services/pharmacy.py
from collections import defaultdict
from datetime import date, timedelta
from typing import List, Optional

from beanie import PydanticObjectId
from pydantic import BaseModel, Field

from app.core.enums import (
    FulfillmentStatus,
    OrderStatus,
    PurchaseOrderStatus,
    ReturnStatus,
    UserRole,
)
from app.core.utils import get_or_404
from app.models.pharmacy import (
    FulfillmentPrescription,
    InventoryItem,
    OrderMedicineItem,
    PharmacyOrder,
    PurchaseOrder,
    PurchaseOrderItem,
    ReturnRequest,
    SupplierInfo,
)
from app.models.user import User


class InventoryOut(BaseModel):
    id: PydanticObjectId = Field(alias="_id")
    medicine_name: str
    batch_number: str
    quantity: int
    expiry_date: str
    supplier: str
    price: float

    model_config = {"populate_by_name": True}


class InventoryCreate(BaseModel):
    medicine_name: str
    batch_number: str
    quantity: int
    expiry_date: str
    supplier: str
    price: float


class OrderCreate(BaseModel):
    medicines: List[OrderMedicineItem]
    address: str
    prescription_id: Optional[PydanticObjectId] = None


class OrderOut(BaseModel):
    id: PydanticObjectId = Field(alias="_id")
    patient_id: PydanticObjectId
    medicines: List[OrderMedicineItem]
    total: float
    status: OrderStatus
    date: str
    address: str

    model_config = {"populate_by_name": True}


class FulfillmentOut(BaseModel):
    id: PydanticObjectId = Field(alias="_id")
    rx_id: str
    prescription_id: PydanticObjectId
    patient_name: str
    doctor_name: str
    date: str
    medicines: List[str]
    status: FulfillmentStatus

    model_config = {"populate_by_name": True}


class PurchaseOrderCreate(BaseModel):
    supplier: str
    items: List[PurchaseOrderItem]
    order_date: str
    expected_delivery: str


class PurchaseOrderOut(BaseModel):
    id: PydanticObjectId = Field(alias="_id")
    supplier: str
    items: List[PurchaseOrderItem]
    total: float
    order_date: str
    expected_delivery: str
    status: PurchaseOrderStatus

    model_config = {"populate_by_name": True}


class ReturnOut(BaseModel):
    id: PydanticObjectId = Field(alias="_id")
    order_id: PydanticObjectId
    patient_name: str
    medicines: List[str]
    amount: float
    reason: str
    request_date: str
    status: ReturnStatus

    model_config = {"populate_by_name": True}


class SupplierOut(BaseModel):
    id: PydanticObjectId = Field(alias="_id")
    name: str
    contact_email: str
    contact_phone: str
    address: str
    products_count: int

    model_config = {"populate_by_name": True}


class ExpiryAlertOut(BaseModel):
    id: PydanticObjectId = Field(alias="_id")
    medicine_name: str
    batch_number: str
    quantity: int
    expiry_date: str
    alert_type: str

    model_config = {"populate_by_name": True}


class SalesSummary(BaseModel):
    monthly_sales: float
    growth_percent: float
    total_orders: int
    units_sold: int
    monthly: List[dict]
    top_sellers: List[dict]


async def list_inventory(low_stock: bool = False) -> List[InventoryOut]:
    items = await InventoryItem.find_all().to_list()
    if low_stock:
        items = [i for i in items if i.quantity < 500]
    return [InventoryOut(_id=i.id, **i.model_dump(exclude={"id"})) for i in items]


async def add_inventory(body: InventoryCreate) -> InventoryOut:
    item = InventoryItem(**body.model_dump())
    await item.insert()
    return InventoryOut(_id=item.id, **item.model_dump(exclude={"id"}))


async def list_orders(current_user: User) -> List[OrderOut]:
    if current_user.role == UserRole.PATIENT:
        items = await PharmacyOrder.find(PharmacyOrder.patient_id == current_user.id).to_list()
    else:
        items = await PharmacyOrder.find_all().to_list()
    return [OrderOut(_id=o.id, **o.model_dump(exclude={"id"})) for o in items]


async def create_order(body: OrderCreate, patient: User) -> OrderOut:
    total = sum(m.quantity * m.price for m in body.medicines)
    order = PharmacyOrder(
        patient_id=patient.id,
        medicines=body.medicines,
        total=total,
        date=str(date.today()),
        address=body.address,
        prescription_id=body.prescription_id,
    )
    await order.insert()
    return OrderOut(_id=order.id, **order.model_dump(exclude={"id"}))


async def update_order_status(order_id: PydanticObjectId, status_value: OrderStatus) -> OrderOut:
    order = await get_or_404(PharmacyOrder, order_id, "Order not found")
    order.status = status_value
    await order.save()
    return OrderOut(_id=order.id, **order.model_dump(exclude={"id"}))


async def list_fulfillment() -> List[FulfillmentOut]:
    items = await FulfillmentPrescription.find_all().to_list()
    return [FulfillmentOut(_id=f.id, **f.model_dump(exclude={"id"})) for f in items]


async def advance_fulfillment(fulfillment_id: PydanticObjectId) -> FulfillmentOut:
    item = await get_or_404(FulfillmentPrescription, fulfillment_id, "Not found")
    flow = [FulfillmentStatus.PENDING, FulfillmentStatus.DISPENSING, FulfillmentStatus.DISPENSED]
    idx = flow.index(item.status)
    if idx < len(flow) - 1:
        item.status = flow[idx + 1]
        await item.save()
    return FulfillmentOut(_id=item.id, **item.model_dump(exclude={"id"}))


async def list_purchase_orders() -> List[PurchaseOrderOut]:
    items = await PurchaseOrder.find_all().to_list()
    return [PurchaseOrderOut(_id=p.id, **p.model_dump(exclude={"id"})) for p in items]


async def create_purchase_order(body: PurchaseOrderCreate) -> PurchaseOrderOut:
    total = sum(i.quantity * i.unit_price for i in body.items)
    po = PurchaseOrder(total=total, status=PurchaseOrderStatus.DRAFT, **body.model_dump())
    await po.insert()
    return PurchaseOrderOut(_id=po.id, **po.model_dump(exclude={"id"}))


async def list_returns() -> List[ReturnOut]:
    items = await ReturnRequest.find_all().to_list()
    return [ReturnOut(_id=r.id, **r.model_dump(exclude={"id", "patient_id"})) for r in items]


async def update_return_status(return_id: PydanticObjectId, status_value: ReturnStatus) -> ReturnOut:
    item = await get_or_404(ReturnRequest, return_id, "Not found")
    item.status = status_value
    await item.save()
    return ReturnOut(_id=item.id, **item.model_dump(exclude={"id", "patient_id"}))


async def list_suppliers() -> List[SupplierOut]:
    items = await SupplierInfo.find_all().to_list()
    return [SupplierOut(_id=s.id, **s.model_dump(exclude={"id"})) for s in items]


async def list_expiry_alerts() -> List[ExpiryAlertOut]:
    items = await InventoryItem.find_all().to_list()
    today = date.today()
    threshold = today + timedelta(days=90)
    alerts: List[ExpiryAlertOut] = []
    for item in items:
        try:
            expiry = date.fromisoformat(item.expiry_date)
        except ValueError:
            continue
        alert_type = None
        if expiry <= today:
            alert_type = "expired"
        elif expiry <= threshold:
            alert_type = "expiring_soon"
        elif item.quantity < 500:
            alert_type = "low_stock"
        if alert_type:
            alerts.append(
                ExpiryAlertOut(
                    _id=item.id,
                    medicine_name=item.medicine_name,
                    batch_number=item.batch_number,
                    quantity=item.quantity,
                    expiry_date=item.expiry_date,
                    alert_type=alert_type,
                )
            )
    return alerts


async def pharmacy_sales() -> SalesSummary:
    orders = await PharmacyOrder.find(PharmacyOrder.status == OrderStatus.DELIVERED).to_list()
    total_revenue = sum(o.total for o in orders)
    units = sum(sum(m.quantity for m in o.medicines) for o in orders)
    monthly_map: dict[str, float] = defaultdict(float)
    for o in orders:
        month_key = o.date[:7] if o.date else "unknown"
        monthly_map[month_key] += o.total
    monthly = [{"month": k, "sales": round(v, 2)} for k, v in sorted(monthly_map.items())[-6:]]
    product_units: dict[str, dict] = defaultdict(lambda: {"units": 0, "revenue": 0.0})
    for o in orders:
        for m in o.medicines:
            product_units[m.name]["units"] += m.quantity
            product_units[m.name]["revenue"] += m.quantity * m.price
    top_sellers = sorted(
        [{"name": k, **v} for k, v in product_units.items()],
        key=lambda x: x["revenue"],
        reverse=True,
    )[:5]
    current = monthly[-1]["sales"] if monthly else 0
    previous = monthly[-2]["sales"] if len(monthly) > 1 else 0
    growth = round(((current - previous) / previous) * 100, 1) if previous else 0
    return SalesSummary(
        monthly_sales=round(current, 2),
        growth_percent=growth,
        total_orders=len(orders),
        units_sold=units,
        monthly=monthly,
        top_sellers=top_sellers,
    )

File: Frontend/api/pharmacy.ts
import { api } from '@/lib/api';
import { mapInventory, mapOrder, BackendInventory, BackendOrder } from '@/lib/mappers';
import { OrderMedicine } from '@/types';

export type { BackendInventory, BackendOrder };
export { mapInventory, mapOrder };

// ---- Inventory ----

export function listInventory(lowStockOnly?: boolean) {
  return api.get<BackendInventory[]>('/pharmacy/inventory', lowStockOnly ? { low_stock: true } : undefined);
}

export interface AddInventoryInput {
  medicine_name: string;
  batch_number: string;
  quantity: number;
  expiry_date: string;
  supplier: string;
  price: number;
}

/** pharmacist/admin. */
export function addInventoryItem(body: AddInventoryInput) {
  return api.post<BackendInventory>('/pharmacy/inventory', body);
}

// ---- Orders ----

/** Scoped server-side: a patient's own orders, or all orders for pharmacist/admin. */
export function listOrders() {
  return api.get<BackendOrder[]>('/pharmacy/orders');
}

export interface CreateOrderInput {
  medicines: OrderMedicine[];
  address: string;
  prescription_id?: string;
}

/** Patient only. */
export function createOrder(body: CreateOrderInput) {
  return api.post<BackendOrder>('/pharmacy/orders', body);
}

/** pharmacist/admin. `status` is sent as a query param per the backend's signature. */
export function updateOrderStatus(id: string, status: string) {
  return api.patch<BackendOrder>(`/pharmacy/orders/${id}/status`, undefined, { status_value: status });
}

// ---- Fulfillment (dispensing workflow) ----

export type FulfillmentStatus = 'pending' | 'dispensing' | 'dispensed';

export interface FulfillmentRx {
  id: string;
  rxId: string;
  patientName: string;
  doctorName: string;
  date: string;
  medicines: string[];
  status: FulfillmentStatus;
}

export interface BackendFulfillment {
  id: string;
  rx_id: string;
  prescription_id: string;
  patient_name: string;
  doctor_name: string;
  date: string;
  medicines: string[];
  status: FulfillmentStatus;
}

export function mapFulfillment(f: BackendFulfillment): FulfillmentRx {
  return {
    id: f.id,
    rxId: f.rx_id,
    patientName: f.patient_name,
    doctorName: f.doctor_name,
    date: f.date,
    medicines: f.medicines,
    status: f.status
  };
}

/** pharmacist only. */
export function listFulfillment() {
  return api.get<BackendFulfillment[]>('/pharmacy/fulfillment');
}

export function advanceFulfillment(id: string) {
  return api.post<BackendFulfillment>(`/pharmacy/fulfillment/${id}/advance`);
}

// ---- Purchase orders ----

export type PurchaseOrderStatus = 'draft' | 'ordered' | 'shipped' | 'received';

export interface PurchaseOrderItem {
  name: string;
  quantity: number;
  unitPrice: number;
}

export interface PurchaseOrder {
  id: string;
  supplier: string;
  items: PurchaseOrderItem[];
  total: number;
  orderDate: string;
  expectedDelivery: string;
  status: PurchaseOrderStatus;
}

export interface BackendPurchaseOrderItem {
  name: string;
  quantity: number;
  unit_price: number;
}

export interface BackendPurchaseOrder {
  id: string;
  supplier: string;
  items: BackendPurchaseOrderItem[];
  total: number;
  order_date: string;
  expected_delivery: string;
  status: PurchaseOrderStatus;
}

export function mapPurchaseOrder(o: BackendPurchaseOrder): PurchaseOrder {
  return {
    id: o.id,
    supplier: o.supplier,
    items: o.items.map((it) => ({ name: it.name, quantity: it.quantity, unitPrice: it.unit_price })),
    total: o.total,
    orderDate: o.order_date,
    expectedDelivery: o.expected_delivery,
    status: o.status
  };
}

/** pharmacist only. */
export function listPurchaseOrders() {
  return api.get<BackendPurchaseOrder[]>('/pharmacy/purchase-orders');
}

export interface CreatePurchaseOrderInput {
  supplier: string;
  items: BackendPurchaseOrderItem[];
  order_date: string;
  expected_delivery: string;
}

export function createPurchaseOrder(body: CreatePurchaseOrderInput) {
  return api.post<BackendPurchaseOrder>('/pharmacy/purchase-orders', body);
}

// ---- Returns ----

export type ReturnStatus = 'pending' | 'approved' | 'refunded' | 'rejected';

export interface ReturnRequest {
  id: string;
  orderId: string;
  patientName: string;
  medicines: string[];
  amount: number;
  reason: string;
  requestDate: string;
  status: ReturnStatus;
}

export interface BackendReturn {
  id: string;
  order_id: string;
  patient_name: string;
  medicines: string[];
  amount: number;
  reason: string;
  request_date: string;
  status: ReturnStatus;
}

export function mapReturn(r: BackendReturn): ReturnRequest {
  return {
    id: r.id,
    orderId: r.order_id,
    patientName: r.patient_name,
    medicines: r.medicines,
    amount: r.amount,
    reason: r.reason,
    requestDate: r.request_date,
    status: r.status
  };
}

/** pharmacist only. */
export function listReturns() {
  return api.get<BackendReturn[]>('/pharmacy/returns');
}

/** `status` is sent as a query param per the backend's signature. */
export function updateReturnStatus(id: string, status: ReturnStatus) {
  return api.patch<BackendReturn>(`/pharmacy/returns/${id}/status`, undefined, { status_value: status });
}

// ---- Suppliers ----

export interface Supplier {
  id: string;
  name: string;
  email: string;
  phone: string;
  address: string;
  productsCount: number;
}

export interface BackendSupplier {
  id: string;
  name: string;
  contact_email: string;
  contact_phone: string;
  address: string;
  products_count: number;
}

export function mapSupplier(s: BackendSupplier): Supplier {
  return {
    id: s.id,
    name: s.name,
    email: s.contact_email,
    phone: s.contact_phone,
    address: s.address,
    productsCount: s.products_count
  };
}

/** pharmacist/admin. */
export function listSuppliers() {
  return api.get<BackendSupplier[]>('/pharmacy/suppliers');
}

// ---- Expiry alerts ----

export type ExpiryAlertType = 'expired' | 'expiring_soon' | 'low_stock';

export interface ExpiryAlert {
  id: string;
  medicineName: string;
  batchNumber: string;
  quantity: number;
  expiryDate: string;
  isLowStock: boolean;
  isExpiring: boolean;
  isExpired: boolean;
}

export interface BackendExpiryAlert {
  id: string;
  medicine_name: string;
  batch_number: string;
  quantity: number;
  expiry_date: string;
  alert_type: ExpiryAlertType;
}

export function mapExpiryAlert(a: BackendExpiryAlert): ExpiryAlert {
  return {
    id: a.id,
    medicineName: a.medicine_name,
    batchNumber: a.batch_number,
    quantity: a.quantity,
    expiryDate: a.expiry_date,
    isLowStock: a.alert_type === 'low_stock',
    isExpiring: a.alert_type === 'expiring_soon' || a.alert_type === 'expired',
    isExpired: a.alert_type === 'expired'
  };
}

/** pharmacist/admin. */
export function listExpiryAlerts() {
  return api.get<BackendExpiryAlert[]>('/pharmacy/expiry-alerts');
}

// ---- Sales summary ----

export interface MonthlySalesPoint {
  month: string;
  sales: number;
}

export interface TopSeller {
  name: string;
  units: number;
  revenue: number;
}

export interface SalesSummary {
  monthly_sales: number;
  growth_percent: number;
  total_orders: number;
  units_sold: number;
  monthly: MonthlySalesPoint[];
  top_sellers: TopSeller[];
}

/** pharmacist/admin. */
export function getSalesSummary() {
  return api.get<SalesSummary>('/pharmacy/sales');
}
